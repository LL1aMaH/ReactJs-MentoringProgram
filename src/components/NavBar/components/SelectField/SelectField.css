.container {
  font-size: 0.875rem;
  display: inline-flex;
  flex-direction: column;
  position: relative;
  width: 100%;
}

.disabledContainer .selectButton {
  color: var(--disabled-color);
  cursor: default;
}

.selectList {
  position: absolute;
  border-radius: 2px;
  list-style: none;
  padding: 0;
  margin: 15px 0 0 0;
  width: 100%;
  z-index: 1;
  top: 100%;
  max-height: 21rem;
  background-color: var(--secondary-background-color);
}

.close {
  display: none;
}

.open {
  display: flex;
  flex-direction: column;
}

.dropDown {
  min-width: 18rem;
  display: flex;
  justify-content: space-between;
  align-items: center;

  & span {
    font-size: 20px;
    color: var(--text-primary-color);
    white-space: nowrap;
  }
}
