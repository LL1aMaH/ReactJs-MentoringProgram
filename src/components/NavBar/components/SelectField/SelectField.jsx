import React, { useState, useCallback, memo, useEffect } from 'react';

import classnames from 'classnames';

import styles from '../../../../../styles/SelectField.module.scss';

import SelectButton from './components/SelectButton/SelectButton';
import SelectOption from './components/SelectOption/SelectOption';
import useDropdown from './hooks/useDropdown/useDropdown';


export const SelectField = memo(function SelectField(props) {
  const {
    options,
    onChange,
    disabled = false,
    className,
    label,
    containerClassName,
    buttonClassName,
    optionClassName,
    optionListClassName,
    selectedOptionClassName,
    labelClassName,
  } = props;

  const { ref, isOpen, toggleDropdown, closeDropdown } = useDropdown();

  const [selectedValue, setSelectedValue] = useState(options[0]);

  useEffect(() => {
    setSelectedValue(selectedValue);
  }, [selectedValue]);

  const handleClick = useCallback(
    (item) => {
      setSelectedValue(item);

      onChange(item);

      closeDropdown();
    },
    [onChange, closeDropdown],
  );

  const handleToggleDropdown = useCallback(() => {
    if (!disabled) {
      toggleDropdown();
    }
  }, [disabled, toggleDropdown]);

  const layoutClassName = classnames(styles.container, className, {
    [styles.disabledContainer]: disabled,
  });

  const listClassName = classnames(styles.selectList, optionListClassName, {
    [styles.open]: isOpen,
    [styles.close]: !isOpen,
  });

  const droptDownClassName = classnames(styles.dropDown, containerClassName);

  return (
    <div className={droptDownClassName}>
      {label && <span className={labelClassName}>{label}</span>}
      <div className={layoutClassName} ref={ref}>
        <SelectButton
          label={selectedValue}
          onClick={handleToggleDropdown}
          isOpen={isOpen}
          buttonClassName={buttonClassName}
        />
        <ul className={listClassName}>
          {options.map((item) => (
            <SelectOption
              onClick={handleClick}
              key={item}
              label={item}
              value={item}
              isSelected={item === selectedValue}
              optionClassName={optionClassName}
              selectedOptionClassName={selectedOptionClassName}
            />
          ))}
        </ul>
      </div>
    </div>
  );
});
