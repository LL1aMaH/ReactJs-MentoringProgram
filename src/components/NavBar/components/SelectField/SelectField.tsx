import React, { useState, useCallback, memo, useEffect } from 'react';

import classnames from 'classnames';

import SelectButton from './components/SelectButton';
import SelectOption from './components/SelectOption';
import useDropdown from './hooks/useDropdown/useDropdown';

import styles from './SelectField.css';

export type TSelectFieldProps<T> = {
  options: Array<string>;
  disabled?: boolean;
  onChange: (value: T) => void;
  className?: string;
  label?: string;
  containerClassName?: string;
  buttonClassName?: string;
  optionClassName?: string;
  optionListClassName?: string;
  selectedOptionClassName?: string;
  labelClassName?: string;
};

function SelectFieldBase<T>(props: TSelectFieldProps<T>): JSX.Element {
  const {
    options,
    onChange,
    disabled = false,
    className,
    label,
    containerClassName,
    buttonClassName,
    optionClassName,
    optionListClassName,
    selectedOptionClassName,
    labelClassName,
  } = props;

  const { ref, isOpen, toggleDropdown, closeDropdown } = useDropdown();

  const [selectedValue, setSelectedValue] = useState(options[0]);

  useEffect(() => {
    setSelectedValue(selectedValue);
  }, [selectedValue]);

  const handleClick = useCallback(
    (item): void => {
      setSelectedValue(item);

      onChange(item);

      closeDropdown();
    },
    [onChange, closeDropdown],
  );

  const handleToggleDropdown = useCallback(() => {
    if (!disabled) {
      toggleDropdown();
    }
  }, [disabled, toggleDropdown]);

  const layoutClassName = classnames(styles.container, className, {
    [styles.disabledContainer]: disabled,
  });

  const listClassName = classnames(styles.selectList, optionListClassName, {
    [styles.open]: isOpen,
    [styles.close]: !isOpen,
  });

  const droptDownClassName = classnames(styles.dropDown, containerClassName);

  return (
    <div className={droptDownClassName}>
      {label && <span className={labelClassName}>{label}</span>}
      <div className={layoutClassName} ref={ref}>
        <SelectButton
          label={selectedValue}
          onClick={handleToggleDropdown}
          isOpen={isOpen}
          buttonClassName={buttonClassName}
        />
        <ul className={listClassName}>
          {options.map((item: string) => (
            <SelectOption
              onClick={handleClick}
              key={item}
              label={item}
              value={item}
              isSelected={item === selectedValue}
              optionClassName={optionClassName}
              selectedOptionClassName={selectedOptionClassName}
            />
          ))}
        </ul>
      </div>
    </div>
  );
}

export const SelectField = memo(SelectFieldBase) as typeof SelectFieldBase;
