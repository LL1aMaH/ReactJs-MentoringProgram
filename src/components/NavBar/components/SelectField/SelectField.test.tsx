import React from 'react';
import renderer from 'react-test-renderer';

import { fireEvent, render, cleanup } from '@testing-library/react';

import { SORT_BY } from 'Consts/index';
import { SelectField } from './SelectField';

describe('SelectField component', () => {
  const getComponent = (
    { options = SORT_BY, onChange = jest.fn(), disabled = false }: any, // eslint-disable-line @typescript-eslint/no-explicit-any
  ) => <SelectField options={options} onChange={onChange} disabled={disabled} />;

  afterEach(cleanup);

  it('renders correctly', () => {
    const componentJSON = renderer.create(getComponent({})).toJSON();
    expect(componentJSON).toMatchSnapshot();
  });

  it('renders open/close dropdown', () => {
    const container = render(getComponent({}));
    const { getByTestId, asFragment } = container;
    const triggerToOpenSelectField: HTMLElement = getByTestId('selectButton');

    // open dropdown
    fireEvent.click(triggerToOpenSelectField);
    expect(asFragment()).toMatchSnapshot();

    // close dropdown
    fireEvent.click(container.container);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders choose option', () => {
    const container = render(getComponent({}));
    const { getByText, getByTestId } = container;
    const selected: HTMLElement = getByTestId('selectButton');

    fireEvent.click(selected);
    const selectedValue = 'FILTER';
    const triggerToOpenSelectField: HTMLElement = getByText(selectedValue);

    fireEvent.click(triggerToOpenSelectField);

    expect(selected.textContent).toBe('FILTER');
  });

  it('does not open dropdown box on disabled component', () => {
    const { getByText, asFragment } = render(getComponent({ disabled: true }));

    fireEvent.click(getByText('FILTER'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('calls change handler by clicking on dropdown item', () => {
    const handleChange = jest.fn();
    const { getByText } = render(getComponent({ onChange: handleChange }));

    fireEvent.click(getByText(SORT_BY[1]));

    expect(handleChange).toHaveBeenCalledWith(SORT_BY[1]);
  });
});
