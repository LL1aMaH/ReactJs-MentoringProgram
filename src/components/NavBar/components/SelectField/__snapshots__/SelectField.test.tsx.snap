// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField component does not open dropdown box on disabled component 1`] = `
<DocumentFragment>
  <div
    class="dropDown"
  >
    <div
      class="container disabledContainer"
    >
      <button
        class="selectButton dropdownClose button"
        data-testid="selectButton"
        tabindex="0"
        type="button"
      >
        RATING
        <svg />
      </button>
      <ul
        class="selectList close"
      >
        <li>
          <button
            class="option"
            type="button"
          >
            RELEASE DATE
          </button>
        </li>
        <li>
          <button
            class="option "
            type="button"
          >
            RATING
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SelectField component renders correctly 1`] = `
<div
  className="dropDown"
>
  <div
    className="container"
  >
    <button
      className="selectButton dropdownClose button"
      data-testid="selectButton"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="button"
    >
      RELEASE DATE
      <svg />
    </button>
    <ul
      className="selectList close"
    >
      <li>
        <button
          className="option "
          onClick={[Function]}
          type="button"
        >
          RELEASE DATE
        </button>
      </li>
      <li>
        <button
          className="option"
          onClick={[Function]}
          type="button"
        >
          RATING
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`SelectField component renders open/close dropdown 1`] = `
<DocumentFragment>
  <div
    class="dropDown"
  >
    <div
      class="container"
    >
      <button
        class="selectButton dropdownOpen button"
        data-testid="selectButton"
        tabindex="0"
        type="button"
      >
        RELEASE DATE
        <svg />
      </button>
      <ul
        class="selectList open"
      >
        <li>
          <button
            class="option "
            type="button"
          >
            RELEASE DATE
          </button>
        </li>
        <li>
          <button
            class="option"
            type="button"
          >
            RATING
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SelectField component renders open/close dropdown 2`] = `
<DocumentFragment>
  <div
    class="dropDown"
  >
    <div
      class="container"
    >
      <button
        class="selectButton dropdownClose button"
        data-testid="selectButton"
        tabindex="0"
        type="button"
      >
        RELEASE DATE
        <svg />
      </button>
      <ul
        class="selectList close"
      >
        <li>
          <button
            class="option "
            type="button"
          >
            RELEASE DATE
          </button>
        </li>
        <li>
          <button
            class="option"
            type="button"
          >
            RATING
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
